### PR TITLE
Fix ndarray indexing bug

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
Fixed https://github.com/apache/incubator-mxnet/issues/16647

@fhieber

### CODE REVIEW

File: src/ndarray/ndarray.cc [hunk 1]
```diff
@@ -243,8 +243,14 @@ NDArray NDArray::MKLDNNDataReshape(const mxnet::TShape &shape) const {
...
   CHECK(!is_none()) << "NDArray is not initialized";
-  CHECK_GE(shape_.Size(), shape.Size())
-    << "NDArray.Reshape: target shape size is larger current shape";
+  if (Imperative::Get()->is_np_shape()) {
+    CHECK_EQ(shape_.Size(), shape.Size())
+        << "NDArray.Reshape: target shape must have the same size as "
+        << "current shape.";
+  } else {
+    CHECK_GE(shape_.Size(), shape.Size())
+        << "NDArray.Reshape: target shape size is larger current shape";
```
Comment (ptrendx) [inline src/ndarray/ndarray.cc]:
```suggestion
        << "NDArray.Reshape: target shape size is larger than current shape";
```

**Comment by ptrendx** [comment]
@szha Can we get this reviewed today?

@reminisce Since the repro in #16647 is small, can we add a test with it to make sure the original issue is (and stays) actually fixed?

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (reminisce) [reply src/ndarray/ndarray.cc]: Done.

**Comment by reminisce** [comment]
@ptrendx Thanks for the review. I added similar test case in test_numpy_ndarray.py, but I have also included the cases from #16647 in test_ndarray.py as you requested.

**Comment by reminisce** [comment]
@PatricZhao @TaoLv  MKLDNN CPP test keeps failing on `FullyConnectedOp` in this PR. Could you guys help to take a look? Thanks.
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-16895/5/pipeline/366

**Comment by TaoLv** [comment]
@reminisce I have created a PR to revert the problematic commit. Sorry for the inconvenience.

**Comment by ptrendx** [comment]
@reminisce could you merge with the latest master to pick up this fix for the cpp test?

**Comment by ptrendx** [comment]
Build failed due to network error - I retriggered it.

**Comment by reminisce** [comment]
Ubuntu 14.04 has problems. It's being under investigation.

**Pull request merged**